DSharpPlus.Core/Rest/PollyRetryPolicy.cs

```diff
@@ -0,0 +1,50 @@
+using System;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+using Polly;
+using Polly.Retry;
+
+namespace DSharpPlus.Core.Rest
+{
+    internal class PollyRetryPolicy
+    {
+        private readonly TimeSpan __one_second = TimeSpan.FromSeconds(1);
+
+        public PollyRetryPolicy(int internalRequeues)
+            => RetryPolicy =
+                Policy.HandleResult<HttpResponseMessage>(response => response.StatusCode == HttpStatusCode.TooManyRequests
+                    && response.Headers.GetValues("X-DSharpPlus-Preemptive-Ratelimit").Any())
+                    .WaitAndRetryAsync(retryCount: internalRequeues,
+                        sleepDurationProvider: (_, responseTask, _) =>
+                        {
+                            HttpResponseMessage message = responseTask.Result;
+
+                            if (message.Headers.GetValues("X-RateLimit-Scope").SingleOrDefault() == "shared")
+                            {
+                                throw new HttpRequestException("Shared ratelimit hit, not retrying request.");
+                            }
+
+                            if (message == default)
```

Comment:
Also unreachable; HttpResponseMessage is a class, and `default(T)` is `null`, so you'd throw an NRE before you reached this.

Reply:
good catch